homeassistant/components/sensor/bart.py

```diff
@@ -0,0 +1,256 @@
+"""
+Support for the BART (Bay Area Rapid Transit) API.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/sensor.bart/
+"""
+import logging
+from datetime import timedelta
+
+from homeassistant.helpers.entity import Entity
+from homeassistant.util import Throttle
+
+_LOGGER = logging.getLogger(__name__)
+
+ICON = 'mdi:train'
+
+# Return cached results if last scan was less then this time ago.
+MIN_TIME_BETWEEN_UPDATES = timedelta(seconds=60)
+
+
+def parse_api_response(raw_xml):
+    """Parse the BART API response."""
+    import xml.etree.ElementTree as etree
+    if isinstance(raw_xml, bytes):
+        parsed_xml = etree.fromstring(
+            raw_xml, parser=etree.XMLParser(
+                encoding='utf-8'))
+    else:
+        parsed_xml = etree.parse(raw_xml)
+    return parsed_xml
+
+
+def get_bart_xml(url):
+    """Helper function to query the BART API."""
+    from urllib.request import urlopen
+    raw_response = urlopen(url)
+    xml = parse_api_response(raw_response)
+    return xml
+
+
+def parse_bart_etd(etd, origin):
+    """Parse time estimate into usable format."""
+    raw_estimates = etd.findall("estimate")
+    estimates = []
+    for estimate in raw_estimates:
+        appending_estimate = dict(
+            (("departure_" + elt.tag, elt.text) for elt in estimate))
+        if appending_estimate["departure_minutes"] == "Leaving":
+            appending_estimate["departure_minutes"] = "0"
+        appending_estimate["departure_destination"] = etd.find(
+            "destination").text
+        appending_estimate["departure_abbreviation"] = etd.find(
+            "abbreviation").text
+        appending_estimate["provided_station"] = origin
+        estimates.append(appending_estimate)
+    return estimates
+
+
+# pylint: disable=too-few-public-methods
+class BARTAPIClient(object):
+    """The Class for accessing the BART API."""
+
+    def __init__(self, api_key="MW9S-E7SL-26DU-VV8V"):
```

Comment:
@balloob Actually, BART provides that API key publicly and suggests using it. See [here](http://www.bart.gov/schedules/developers/api) (specifically the section titled "Two ways to get a validation key")